lib/Dialect/FIRRTL/Transforms/GrandCentral.cpp

```diff
@@ -0,0 +1,499 @@
+                       ElementInfo::String)
+                 .Case("sifive.enterprise.grandcentral.AugmentedBooleanType",
+                       ElementInfo::Boolean)
+                 .Case("sifive.enterprise.grandcentral.AugmentedIntegerType",
+                       ElementInfo::Integer)
+                 .Case("sifive.enterprise.grandcentral.AugmentedDoubleType",
+                       ElementInfo::Double)
+                 .Default(ElementInfo::Error);
+  if (tpe == ElementInfo::Error)
+    return {};
+
+  StringRef description = {};
+  if (auto maybeDescription = field.getAs<StringAttr>("description"))
+    description = maybeDescription.getValue();
+  return Optional<AugmentedField>({name.getValue(), description, tpe});
+};
+
+/// Convert an Annotation into an optional AugmentedBundleType.  Returns None if
+/// the annotation is not an AugmentedBundleType.
+static Optional<AugmentedBundleType> decodeBundleType(Annotation anno) {
+  auto defName = anno.getMember<StringAttr>("defName");
+  auto elements = anno.getMember<ArrayAttr>("elements");
+  if (!defName || !elements)
+    return {};
+  auto bundle =
+      AugmentedBundleType({defName.getValue(), SmallVector<AugmentedField>()});
+  for (auto element : elements) {
+    auto field = decodeField(element);
+    if (!field)
+      return {};
+    bundle.elements.push_back(field.getValue());
+  }
+  return Optional<AugmentedBundleType>(bundle);
+};
+
+/// Remove Grand Central Annotations associated with SystemVerilog interfaces
+/// that should emitted.  This pass works in three major phases:
+///
+/// 1. The circuit's annotations are examnined to figure out _what_ interfaces
+///    there are.  This includes information about the name of the interface
+///    ("defName") and each of the elements (sv::InterfaceSignalOp) that make up
+///    the interface.  However, no information about the _type_ of the elements
+///    is known.
+///
+/// 2. With this, information, walk through the circuit to find scattered
+///    information about the types of the interface elements.  Annotations are
+///    scattered during FIRRTL parsing to attach all the annotations associated
+///    with elements on the right components.
+///
+/// 3. Add interface ops and populate the elements.
+///
+/// Grand Central supports three "normal" element types and four "weird" element
+/// types.  The normal ones are ground types (SystemVerilog logic), vector types
+/// (SystemVerilog unpacked arrays), and nested interface types (another
+/// SystemVerilog interface).  The Chisel API provides "weird" elements that
+/// include: Boolean, Integer, String, and Double.  The SFC implementation
+/// currently drops these, but this pass emits them as commented out strings.
+struct GrandCentralPass : public GrandCentralBase<GrandCentralPass> {
+  void runOnOperation() override;
+
+private:
+  // Discovered interfaces that need to be constructed.
+  llvm::DenseMap<std::pair<StringRef, StringRef>, ElementInfo> interfaceMap;
```

Comment:
Great catch and thanks for the pointer with the explanation. Moving this into a local variable makes sense to me.

Reply:
There's an argument to be made I think for keeping this kind of stuff with the pass itself, in that if it's run on a lot of operations, you would end up re-allocating storage less and less, as you're subequent ops are likely to require a similar amount of heap storage.

Running on a `CircuitOp` that doesn't really apply here. :+1: for moving into a local variable.

Reply:
Preventing the reallocation of that (and the other big datastructures here like `interfaces` and `interfaceKeys`) makes me inclined to move those into private class variables and then clear them. 

(I was curious if I would need to _then_ start clearing these datastructures on error. However, it looks like `signalPassFailure()` guarantees that no other passes will run so that mutable state doesn't matter. 😌)

Reply:
Will switch to explicit clears.